### PR TITLE
(QENG-4823) AWS EC2 Reaper

### PR DESCRIPTION
**Pull request by tvpartytonight**
This is the initial commit for the AWS EC2 Reaper.

### CODE REVIEW

**Comment by tvpartytonight** [comment]
@binford2k @doug-rosser I'm going to have the QE team review this, but I just wanted to get more eyes on this at this stage.

**Comment by tvpartytonight** [comment]
This is a mostly done work in progress; the scripts themselves have the destructive parts commented out at the moment, so running them in ec2 is essentially a no-op.

File: lambdas/ec2/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  }
+}
+```
+
+The Terminator should run on a Cloudwatch schedule, configurable in the AWS EC2 GUI. 
+
+## Components
+
+### Schema Enforcer
+The Schema Enforcer is a lambda that is designed to be triggered when an EC2 instance enters the 'pending' state. This lambda waits for an ec2 instance to have a valid `lifetime` tag associated with it. When that tag is found, it converts that tag to a UTC date and adds a `termination_date` tag to the instance with that UTC date as its value. The Schema Enforcer terminates instances that do no have valid `termination_date` tags after 4 minutes, and reports that termination as an exception raised. The Schema Enforcer does not terminate instances past their `termination_date`; the Terminator is responsible for that.
```
Comment (doug-rosser) [inline lambdas/ec2/README.md]: This isn't 100% clear. If a host doesn't have a "lifetime" tag, it does or does not get a "termination_date" tag?

File: lambdas/ec2/schema_enforcer.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                        'Value': termination_date.isoformat()
+                    }
+                ]
+            )
+    else:
+        print('No termination_date found; terminating instance')
+        # TODO: This code should also eventually report terminated instances
+        # to some log somewhere.
+        # ec2_instance.terminate()
+        raise Exception('EC2 instance deleted because no termination_date found within 4 minutes of creation.')
```
Comment (doug-rosser) [inline lambdas/ec2/schema_enforcer.py]: Sort of "smelly" that the enforcer also terminates (even though it doesn't).

File: lambdas/ec2/schema_enforcer.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    if unit == 'w':
+        return datetime.timedelta(weeks=length)
+    elif unit =='h':
+        return datetime.timedelta(hours=length)
+    elif unit == 'd':
+        return datetime.timedelta(days=length)
+    else:
+        raise ValueError("Unable to parse the unit '{0}'".format(unit))
+
+ec2 = boto3.resource('ec2')
```
Comment (doug-rosser) [inline lambdas/ec2/schema_enforcer.py]: What do here?

File: lambdas/ec2/README.md (hunk 1 above)
Comment (tvpartytonight) [reply lambdas/ec2/README.md]: I think this description is missing a portion that is in the "Usage" section above:

>  Users can bypass this workflow with the `lifetime` key and set the `termination_date` directly

If a host doesn't have a `lifetime` tag, the schema enforcer will never process that tag and convert it into a 'termination_date' tag. However, you can supply a `termination_date` tag directly, and it will just process that `termination_date` tag without the lifetime conversion.

File: lambdas/ec2/schema_enforcer.py (hunk 2 above)
Comment (tvpartytonight) [reply lambdas/ec2/schema_enforcer.py]: Yeah, the Enforcer deletes it if it doesn't comply with the rules. I think this confusion points to the fact that I should have a Rules section as well, to really spell it out.

File: lambdas/ec2/schema_enforcer.py (hunk 3 above)
Comment (tvpartytonight) [reply lambdas/ec2/schema_enforcer.py]: This is the assignment of the boto3 abstraction for ec2; I wasn't sure if there was functionality that should exist with this outside of the lambda handler, so I left it here in the global scope of the script in case it was needed outside of the handler.

**Comment by doug-rosser** [comment]
I'm 👍  except that a "safety" should be implemented to turn termination off/on globally.

File: lambdas/ec2/schema_enforcer.py [hunk 4]
```diff
@@ -6,26 +6,35 @@
...
 
-    This method returns False if the ec2 instance currently has no tags
+    This method returns None if the ec2 instance currently has no tags
     or if the tag is not found. If the tag is found, it returns the tag
     value.
     """
-    ec2_instance.load()
     if ec2_instance.tags is None:
-        return False
+        return None
```
Comment (doug-rosser) [inline lambdas/ec2/schema_enforcer.py:34]: Why the change here? It scares me.

`>>> q = None
>>> if q is False:
...   print "None is False"
... else:
...   print "None is NOT False"
... 
None is NOT False`

File: lambdas/ec2/schema_enforcer.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        raise
+
+def validate_ec2_termination_date(ec2_instance):
+    """
+    :param ec2_instance: a boto3 resource representing an Amazon EC2 Instance.
+
+    Validates that an ec2 instance has a valid termination_date in the future.
+    """
+    ec2_termination_date = get_tag(ec2_instance, 'termination_date')
+    if ec2_termination_date is False:
```
Comment (doug-rosser) [inline lambdas/ec2/schema_enforcer.py]: See my previous comment where None does not appear to be (is) False.

Comment (tvpartytonight) [reply lambdas/ec2/schema_enforcer.py]: This is a bug here, it should be `None` instead of `False`.

File: lambdas/ec2/schema_enforcer.py (hunk 4 above)
Comment (tvpartytonight) [reply lambdas/ec2/schema_enforcer.py:34]: I changed it here to reflect more accurately that the tag did not exist, not that it was a False value.

In your example, `None` is not `False`, so that is why that logic evaluates like that. Here is a more accurate snippet of `None` evaluates with `if`:

````
>>> if None:
...     print('None is True!')
... else:
...     print('None is False!')
None is False!
````

**Comment by doug-rosser** [comment]
I am officially 👍
